distributed/scheduler.py

```diff
@@ -6304,6 +6305,17 @@ async def register_worker_plugin(self, comm, plugin, name=None):
         )
         return responses
 
+    async def unregister_worker_plugin(self, comm, name):
+        """ Unregisters a worker plugin"""
+
+        for idx, entry in enumerate(self.worker_plugins):
+            if entry["name"] == name:
+                break
```

Comment:
We might want to add an `else` condition here (to make this a [`for`/`else` loop](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops)) so we can raise an informative error if `self.worker_plugins` doesn't contain an entry for a plugin with the name `name`

Reply:
Alternatively, we might also want to think about updating `self.worker_plugins` to be a dictionary (similar to `Worker.plugins`) so we could just simply `pop` the desired plugin off `self.worker_plugins`. Though that's a larger change, which may have additional knock off effects

Reply:
Added an else to the for/else loop to raise a `keyError` with a more descriptive message.

Reply:
Regarding updating the `self.worker_plugins` to be a dictionary I'm having an error when trying to do so. I pushed some comments with the changes I would do to make this work. But when trying to run the test that registers a plugin `test_create_with_client` it throws an error in `worker.py` line 926.

Reply:
Thanks for pushing up those changes, they're definitely heading in the right direction! In addition to making changes to the scheduler, we'll also need to make sure that the workers can handle this message

https://github.com/dask/distributed/blob/8dfaca632403b14bf73350780201394d1dd0e5ba/distributed/scheduler.py#L4004-L4009

that the scheduler sends to a new worker when it's added to the cluster. Specifically here

https://github.com/dask/distributed/blob/8dfaca632403b14bf73350780201394d1dd0e5ba/distributed/worker.py#L922-L927

is where the worker processes the `"worker-plugins"` part of this message. We'll need make an update like the following since `response["worker-plugins"]` is now a `dict` instead of a `list`:

```diff
diff --git a/distributed/worker.py b/distributed/worker.py
index 413425b5..05ba5c41 100644
--- a/distributed/worker.py
+++ b/distributed/worker.py
@@ -922,8 +922,8 @@ class Worker(ServerNode):
         else:
             await asyncio.gather(
                 *[
-                    self.plugin_add(**plugin_kwargs)
-                    for plugin_kwargs in response["worker-plugins"]
+                    self.plugin_add(name=name, plugin=plugin)
+                    for name, plugin in response["worker-plugins"].items()
                 ]
             )
```

Reply:
Thanks @jrbourbeau. I modify this and the tweak the test to work properly, now all test are green after changing the data structure of `worker_plugins` to a `dict()`.